session/session.go

```diff
@@ -2621,22 +2622,298 @@ func (s *session) Auth(user *auth.UserIdentity, authentication, salt []byte) err
+		d := time.Now().Unix() - lastChanged
+		if d > lockTime*24*60*60 {
+			// Generate unlock json string.
+			plJson = pm.BuildPasswordLockingJSON(pl.FailedLoginAttempts,
+				pl.PasswordLockTimeDays, "N", 0, time.Now().Format(time.UnixDate))
+		} else {
+			lds := strconv.FormatInt(lockTime, 10)
+			rds := strconv.FormatInt(int64(math.Ceil(float64(lockTime)-float64(d)/(24*60*60))), 10)
+			errRB := failedLoginTrackingRollback(s)
+			if errRB != nil {
+				return errRB
+			}
+			return privileges.GenerateAccountAutoLockErr(pl.FailedLoginAttempts, user, host, lds, rds)
+		}
+	}
+	if plJson != "" {
+		if err = s.passwordLocking(user, host, plJson); err != nil {
+			errRB := failedLoginTrackingRollback(s)
+			if errRB != nil {
+				return errRB
+			}
+			return err
+		} else {
+			err = failedLoginTrackingCommit(s)
+			if err != nil {
+				return err
+			}
+			lockStatusChanged = true
+		}
+	}
+	if lockStatusChanged {
+		return domain.GetDomain(s).NotifyUpdatePrivilege()
+	}
+	return nil
+}
+
+func authFailedTracking(s *session, user string, host string) error {
+	err := failedLoginTrackingBegin(s)
+	if err != nil {
+		return err
+	}
+	// Obtain the number of consecutive password login failures.
+	passwordLocking, getErr := getFailedLoginUserAttributes(s, user, host)
+	if getErr != nil {
+		if rollBackErr := failedLoginTrackingRollback(s); rollBackErr != nil {
+			return rollBackErr
+		}
+		return getErr
+	}
+	// Consecutive wrong password login failure times +1,
+	// If the lock condition is satisfied, the lock status is updated and the update cache is notified.
+	lockStatusChanged, err := userAutoAccountLocked(s, user, host, passwordLocking)
+	if err != nil {
+		if rollBackErr := failedLoginTrackingRollback(s); rollBackErr != nil {
+			return rollBackErr
+		}
+		return err
+	}
+	if commitErr := failedLoginTrackingCommit(s); commitErr != nil {
+		return commitErr
+	}
+	if lockStatusChanged {
+		// Don't want to update the cache frequently, and only trigger the update cache when the lock status is updated.
+		return domain.GetDomain(s).NotifyUpdatePrivilege()
+	}
+	return nil
+}
+
+func (s *session) passwordLocking(user string, host string, newAttributesStr string) error {
+	sql := new(strings.Builder)
+	sqlexec.MustFormatSQL(sql, "UPDATE %n.%n SET ", mysql.SystemDB, mysql.UserTable)
+	sqlexec.MustFormatSQL(sql, "user_attributes=json_merge_patch(coalesce(user_attributes, '{}'), %?)", newAttributesStr)
+	sqlexec.MustFormatSQL(sql, " WHERE Host=%? and User=%?;", host, user)
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnPrivilege)
+	_, err := s.ExecuteInternal(ctx, sql.String())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func failedLoginTrackingBegin(s *session) error {
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnPrivilege)
+	_, err := s.ExecuteInternal(ctx, "BEGIN PESSIMISTIC")
+	if err != nil {
+		if rollBackErr := failedLoginTrackingRollback(s); rollBackErr != nil {
```

Comment:
You don't need to make a 'ROLLBACK' here, since the begin of transaction failed.